commit 7b69cd2f6c28a088c886a51549005b8edcbb91f1
Author: Eike Hein <hein@kde.org>
Date:   Wed Aug 14 16:21:39 2013 +0200

    Make the setData() accessor take a QVariant as payload.
    
    The implementation now specifically tries to support byte arrays
    being passed in, which is what the base class speaks. To try and
    retain the previous behavior anything else is attempted to be
    converted to a string and failing that, silently dropped.
    
    Backport from plasma-framework f63357bd and reviewed in 111748.
    
    CCBUG:322567

diff --git a/plasma/declarativeimports/draganddrop/DeclarativeMimeData.cpp b/plasma/declarativeimports/draganddrop/DeclarativeMimeData.cpp
index 28532a2..c3857cb 100644
--- a/plasma/declarativeimports/draganddrop/DeclarativeMimeData.cpp
+++ b/plasma/declarativeimports/draganddrop/DeclarativeMimeData.cpp
@@ -118,9 +118,13 @@ void DeclarativeMimeData::setColor(const QColor &color)
     }
 }
 
-void DeclarativeMimeData::setData(const QString &mimeType, const QString &data)
+void DeclarativeMimeData::setData(const QString &mimeType, const QVariant &data)
 {
-    QMimeData::setData(mimeType, data.toLatin1());
+    if (data.type() == QVariant::ByteArray) {
+        QMimeData::setData(mimeType, data.toByteArray());
+    } else if (data.canConvert(QVariant::String)) {
+        QMimeData::setData(mimeType, data.toString().toLatin1());
+    }
 }
 
 /*!
diff --git a/plasma/declarativeimports/draganddrop/DeclarativeMimeData.h b/plasma/declarativeimports/draganddrop/DeclarativeMimeData.h
index 8342938..1faf17e 100644
--- a/plasma/declarativeimports/draganddrop/DeclarativeMimeData.h
+++ b/plasma/declarativeimports/draganddrop/DeclarativeMimeData.h
@@ -78,7 +78,7 @@ public:
     QColor color() const;
     void setColor(const QColor &color);
 
-    Q_INVOKABLE void setData(const QString &mimeType, const QString &data);
+    Q_INVOKABLE void setData(const QString &mimeType, const QVariant &data);
 
     QDeclarativeItem* source() const;
     void setSource(QDeclarativeItem* source);
